File: 3_examples/java/src/main/java/no/systek/graphqlworkshop/clients/MarketPriceClient.java
package no.systek.graphqlworkshop.clients;

import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Component
public class MarketPriceClient {

  private WebClient client = WebClient.create("https://market.gql.systek.dev/price");

  public double getMarketPrice(String ingredient) {
    MarketPriceResponse marketPriceResponse = client.get()
        .uri(uriBuilder ->
            uriBuilder.queryParam("ingredient", ingredient.toLowerCase()).build())
        .accept(MediaType.APPLICATION_JSON)
        .retrieve()
        .bodyToMono(MarketPriceResponse.class)
        .block();

    return marketPriceResponse == null ? -1 : marketPriceResponse.getPrice();
  }
}

File: 3_examples/kotlin/src/main/kotlin/no/systek/graphqlworkshop/clients/MarketPriceClient.kt
package no.systek.graphqlworkshop.clients

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody

@Component
class MarketPriceClient {
    private val client = WebClient.create("https://market.gql.systek.dev")

    suspend fun getMarketPrice(ingredient: String): MarketPriceResponse =
        client.get()
            .uri {
                it.path("/price")
                    .queryParam("ingredient", ingredient.toLowerCase())
                    .build()
            }
            .also { println("Requesting price for $ingredient") }
            .retrieve()
            .awaitBody()
}
